Enhance the functionality of the `bump_version.sh` script

mcdonnnj opened the pull request:
# <!-- Use the title to describe PR changes in the imperative mood --> #

## 🗣 Description ##

This pull request adds some functionality to the `bump_version.sh` script and DRYs out some of the logic.

> [!Note]
> If/when this pull request is approved I plan on porting the changes over to other skeletons that have a `bump_version.sh` script.
<!-- Describe the "what" of your changes in detail. -->
<!-- To avoid scope creep, limit changes to a single goal. -->

## 💭 Motivation and context ##

I have used custom tokens for prerelease versions in some of my projects (`dev` instead of `rc` as an example) to better articulate what the version represents. I thought about it and determined that building the functionality into the script would be useful. While I was making the changes to support custom tokens I realized it would also make sense to do some general cleanup of the script.
<!-- Why is this change required? -->
<!-- What problem does this change solve? How did you solve it? -->
<!-- Mention any related issue(s) here using appropriate keywords such -->
<!-- as "closes" or "resolves" to auto-close them on merge. -->

## 🧪 Testing ##

Automated tests pass. I verified that the script works as expected both with and without custom tokens for `build`/`prerelease` actions. I also verified that other actions worked as expected.
<!-- How did you test your changes? How could someone else test this PR? -->
<!-- Include details of your testing environment, and the tests you ran to -->
<!-- see how your change affects other areas of the code, etc. -->

<!--
## 📷 Screenshots (if appropriate) ##

Uncomment this section if a screenshot is needed.

-->

## ✅ Pre-approval checklist ##

<!-- Remove any of the following that do not apply. -->
<!-- Draft PRs should have one or more unchecked boxes. -->
<!-- If you're unsure about any of these, don't hesitate to ask. -->
<!-- We're here to help! -->

- [x] This PR has an informative and human-readable title.
- [x] Changes are limited to a single goal - *eschew scope creep!*
- [x] All relevant type-of-change labels have been added.
- [x] I have read the [CONTRIBUTING](../blob/develop/CONTRIBUTING.md) document.
- [x] These code changes follow [cisagov code standards](https://github.com/cisagov/development-guide).
- [x] All new and existing tests pass.

mcdonnnj committed:
Add the ability to specify a token in bump_version.sh

The token is optionally used for the build and prerelease actions to
override the default values used by the semver Python package.

mcdonnnj committed:
DRY out logic to update version information

Since the replacement and commit of version change information is
almost identical across the different logic flows it makes sense to DRY
out the code.

mcdonnnj committed:
Change when existing version information is retrieved

Only retrieve the existing version information if the input to the
bump_version.sh script isn't immediately invalid (if the script does
not get 1-2 arguments).

dav3r commented on bump_version.sh:
```diff
@@ -9,46 +15,68 @@ set -o pipefail
+      ;;
+    build | prerelease)
+      if [ $# -eq 2 ]; then
+        new_version=$(python -c "import semver; print(semver.bump_$1('$old_version', token='$2'))")
```
If the provided token doesn't match the existing token, it looks `semver` doesn't modify the version or throw an error:
```
>>> import semver
>>> version_info = semver.VersionInfo.parse("1.2.3-rc.4")
>>> version_info.bump_prerelease()
Version(major=1, minor=2, patch=3, prerelease='rc.5', build=None)
>>> version_info.bump_prerelease(token="foo")
Version(major=1, minor=2, patch=3, prerelease='rc.5', build=None)
```

Is that the desired behavior here?  Can you please share output of a manual run of this code in this case so that we can see how the script behaves?

mcdonnnj replied:
You are not mirroring how this is done in the script in your example. The note in the help section clarifies that the provided token is not used if it does not match, but it will still version bump:

```console
>>> import semver
>>> old_version = "1.2.3-rc.4"
>>> semver.bump_prerelease(old_version)
'1.2.3-rc.5'
>>> semver.bump_prerelease(old_version, token="test")
'1.2.3-rc.5'
>>> old_version = "1.2.3"
>>> semver.bump_prerelease(old_version)
'1.2.3-rc.1'
>>> semver.bump_prerelease(old_version, token="test")
'1.2.3-test.1'
```

Edit:
After looking at your example it is actually doing things correctly. The `bump_prerelease()` method returns a _new_ `Version` object:

```console
>>> import semver
>>> version_info = semver.VersionInfo.parse("1.2.3-rc.4")
>>> version_info.bump_prerelease()
Version(major=1, minor=2, patch=3, prerelease='rc.5', build=None)
>>> version_info.bump_prerelease()
Version(major=1, minor=2, patch=3, prerelease='rc.5', build=None)
```

So it is still incrementing the version but using the old token. This is the expected, and documented, behavior.

dav3r replied:
Alright, I'm tracking now.  I was a little confused before, but this all makes sense.

I don't totally love the silent no-op (not exactly the right phrase, since it does bump the version, but it ignores the token) if you attempt to use a different token than the current version, but your note in the usage text does indeed mention this behavior.

mcdonnnj replied:
Yes I'm not a fan of it silently ignoring the provided token, but the Python logic completely ignores the token if the version string is already a prerelease per the [sourec code](https://github.com/python-semver/python-semver/blob/8fe4ef678e1cbacd904fbd9f67a8cb3f971642b0/src/semver/version.py#L326-L333). Making sure to explicitly mention/document the behavior was the only thing I could see doing since I'd rather expose the functionality even with the caveat.

dav3r reviewed (commented):
This looks like a nice improvement, but I want to get clarification on one thing before I approve.

dav3r reviewed (approved):
👍 👍

jsf9k commented on bump_version.sh:
```diff
@@ -9,46 +15,68 @@ set -o pipefail
+  mv $tmp_file $README_FILE
+  git add $VERSION_FILE $README_FILE
+  git commit --message "$3"
+  git push
```
I have been thinking for a while that this `git push` should be left to the user when he or she is ready.  Removing it will cut down on needless GitHub Actions runs, which is better for the environment.
```suggestion
```

mcdonnnj replied:
I am completely on board with that (and it has been a peeve of mine for a while). @dav3r any concerns?

dav3r replied:
None- I wholeheartedly agree.

jsf9k reviewed (approved):
I made one suggestion, which need not impede the approval of this pull request.

mcdonnnj committed:
Remove git push from `bump_version.sh` script

The team has come to the consensus that the `git push` should be at the
discretion of the user. This both prevents needless GitHub Actions runs
and ensures that the repository is updated when the user feels it is in
an appropriate state.

Co-authored-by: Shane Frasier <jeremy.frasier@gwe.cisa.dhs.gov>
